src/topology/covering.lean

```diff
@@ -0,0 +1,133 @@
+discrete_topology I ∧ ∃ t : trivialization I f, x ∈ t.base_set
+
+namespace is_evenly_covered
+
+variables {f}
+
+/-- If `x` is evenly covered by `f`, then we can construct a trivialization of `f` at `x`. -/
+noncomputable def to_trivialization {x : X} {I : Type*} [topological_space I]
+  (h : is_evenly_covered f x I) : trivialization (f ⁻¹' {x}) f :=
+(classical.some h.2).trans_fiber_homeomorph ((classical.some h.2).preimage_singleton_homeomorph
+  (classical.some_spec h.2)).symm
+
+lemma mem_to_trivialization_base_set {x : X} {I : Type*} [topological_space I]
+  (h : is_evenly_covered f x I) : x ∈ h.to_trivialization.base_set :=
+classical.some_spec h.2
+
+lemma to_trivialization_apply {x : E} {I : Type*} [topological_space I]
+  (h : is_evenly_covered f (f x) I) : (h.to_trivialization x).2 = ⟨x, rfl⟩ :=
+let e := classical.some h.2, h := classical.some_spec h.2, he := e.mk_proj_snd' h in
+  subtype.ext ((e.to_local_equiv.eq_symm_apply (e.mem_source.mpr h)
+    (by rwa [he, e.mem_target, e.coe_fst (e.mem_source.mpr h)])).mpr he.symm).symm
+
+lemma continuous_at {x : E} {I : Type*} [topological_space I]
+  (h : is_evenly_covered f (f x) I) : continuous_at f x :=
+let e := h.to_trivialization in
+  e.continuous_at_proj (e.mem_source.mpr (mem_to_trivialization_base_set h))
+
+lemma to_is_evenly_covered_preimage {x : X} {I : Type*} [topological_space I]
+  (h : is_evenly_covered f x I) : is_evenly_covered f x (f ⁻¹' {x}) :=
+let ⟨h1, h2⟩ := h in by exactI ⟨((classical.some h2).preimage_singleton_homeomorph
+  (classical.some_spec h2)).embedding.discrete_topology, _, h.mem_to_trivialization_base_set⟩
+
+end is_evenly_covered
+
+/-- A covering map is a continuous function `f : E → X` with discrete fibers such that each point
+  of `X` has an evenly covered neighborhood. -/
+def is_covering_map_on :=
+∀ x ∈ s, is_evenly_covered f x (f ⁻¹' {x})
+
+namespace is_covering_map_on
+
+lemma mk (F : X → Type*) [Π x, topological_space (F x)] [hF : Π x, discrete_topology (F x)]
+  (e : Π x ∈ s, trivialization (F x) f) (h : ∀ (x : X) (hx : x ∈ s), x ∈ (e x hx).base_set) :
+  is_covering_map_on f s :=
+λ x hx, is_evenly_covered.to_is_evenly_covered_preimage ⟨hF x, e x hx, h x hx⟩
+
+end is_covering_map_on
+
+/-- A covering map is a continuous function `f : E → X` with discrete fibers such that each point
+  of `X` has an evenly covered neighborhood. -/
+def is_covering_map :=
+∀ x, is_evenly_covered f x (f ⁻¹' {x})
+
+namespace is_covering_map
+
+lemma mk (F : X → Type*) [Π x, topological_space (F x)] [hF : Π x, discrete_topology (F x)]
+  (e : Π x, trivialization (F x) f) (h : ∀ x, x ∈ (e x).base_set) : is_covering_map f :=
+λ x, is_evenly_covered.to_is_evenly_covered_preimage ⟨hF x, e x, h x⟩
+
+variables {f}
+
+lemma continuous (hf : is_covering_map f) : continuous f :=
+continuous_iff_continuous_at.mpr (λ x, (hf (f x)).continuous_at)
+
+lemma is_locally_homeomorph (hf : is_covering_map f) : is_locally_homeomorph f :=
```

Comment:
Optionally: I think it would be nice to generalise this to `is_covering_map_on` (in the case when `s` is open).

Reply:
I'll have to think about this. It's a tad bit tricker than it looks, since you don't know that `f ⁻¹' s` is open.

Reply:
> you don't know that `f ⁻¹' s` is open.

I don't understand: `f` is continuous and `s` is open so this is trivial, right?

Thinking about this a bit more, I suppose you'll also need that `s` is non-empty but otherwise this should be easy.

Reply:
Oh, are you talking about `is_locally_homeomorph`? I was talking about `continuous`.

Reply:
Is this along the lines of what you were thinking of? #17114